src/Server/TCPHandler.cpp

```diff
@@ -1193,6 +1194,17 @@ void TCPHandler::sendHello()
         writeStringBinary(server_display_name, *out);
     if (client_tcp_protocol_version >= DBMS_MIN_REVISION_WITH_VERSION_PATCH)
         writeVarUInt(DBMS_VERSION_PATCH, *out);
+    if (client_tcp_protocol_version >= DBMS_MIN_PROTOCOL_VERSION_WITH_PASSWORD_COMPLEXITY_RULES)
+    {
+        auto rules = server.context()->getAccessControl().getPasswordComplexityRules();
+
+        writeVarUInt(rules.size(), *out);
+        for (const auto & [original_pattern, exception_message] : rules)
+        {
+            writeStringBinary(original_pattern, *out);
+            writeStringBinary(exception_message, *out);
+        }
+    }
```

Comment:
So the password complexity rules are going to work only for TCP protocol?

Reply:
It seems like we only need to add complexity rules handling to clickhouse-client and therefore to our native TCP protocol. In all other cases (at least as I'm aware) we send the original password to the server, and therefore the server will check passwords without any problems.

Reply:
I got it. A test for http protocol would be useful